### PR TITLE
Assignable VRAM bank allocations

### PR DESCRIPTION
**Pull request by DiscostewSM**
This change integrates being able to allocate to a specific VRAM bank
into the existing implementation.

### CODE REVIEW

**Comment by DiscostewSM** [comment]
Whoops, made a few mistakes. Let me work on it.

**Comment by fincs** [comment]
What is the point of segregating VRAM allocations according to bank; is there anything in the 3DS hardware or operating system that specifically requires VRAM_A or VRAM_B addresses?

**Comment by DiscostewSM** [comment]
This is mostly because of things I have read, but if I recall, there are a few issues with defining it as one big bank. One is that certain texture formats, if not all, cannot span over both banks. Another is with multi-texturing. While I think it's recommended to have textures in the same bank when doing multi-texturing (for performance benefit), the system supposedly becomes unstable if multi-texturing of 4-bit formats and non-4-bit formats together from the same VRAM bank.

Other than that, I had assumed that because of having multiple banks would allow reading from one and writing to the other for a possible performance boost. Whether or not that's true, I don't know.

File: libctru/include/3ds/allocator/vram.h [hunk 1]
```diff
@@ -4,6 +4,21 @@
  */
 #pragma once
 
+typedef enum
+{
+	VRAM_A  = 0x00,
+	VRAM_B  = 0x01,
+	VRAM_AB = 0x02
+} VRAM_ALLOCATOR;
```
Comment (mtheall) [inline libctru/include/3ds/allocator/vram.h]: how about:

    VRAM_A = BIT(0),
    VRAM_B = BIT(1),
    VRAM_AB = VRAM_A | VRAM_B,

File: libctru/source/allocator/vram.cpp [hunk 2]
```diff
@@ -8,21 +8,28 @@ extern "C"
...
 	{
-		sVramPool.AddBlock(blk);
-		rbtree_init(&sAddrMap, addrMapNodeComparator);
-		return true;
+		auto blkB = MemBlock::Create((u8*)0x1F300000, 0x00300000);
+		if(blkB)
+		{
+			sVramPoolA.AddBlock(blkA);
+			sVramPoolB.AddBlock(blkB);
+			rbtree_init(&sAddrMap, addrMapNodeComparator);
```
Comment (mtheall) [inline libctru/source/allocator/vram.cpp]: Do you want two separate address maps?

File: libctru/source/allocator/vram.cpp [hunk 3]
```diff
@@ -39,27 +46,65 @@ void* vramMemAlign(size_t size, size_t alignment)
...
+	}
 
 	auto node = newNode(chunk);
 	if (!node)
 	{
-		sVramPool.Deallocate(chunk);
+		if(bankSet)
+			sVramPoolB.Deallocate(chunk);
+		else
+			sVramPoolA.Deallocate(chunk);
```
Comment (mtheall) [inline libctru/source/allocator/vram.cpp]: if(bank & VRAM_A)
        sVramPoolA.Deallocate(chunk);
    else if(bank & VRAM_B)
        sVramPoolB.Deallocate(chunk);

File: libctru/source/allocator/vram.cpp [hunk 4]
```diff
@@ -39,27 +46,65 @@ void* vramMemAlign(size_t size, size_t alignment)
...
+		else
+			bankSet = 0;
+	}
+	else
+	{
+		// Attempt Bank B only
+		if(!sVramPoolB.Allocate(chunk, size, shift))
+			return nullptr;
+		bankSet = 1;
+	}
```
Comment (mtheall) [inline libctru/source/allocator/vram.cpp]: if(bank & VRAM_A)
    {
        if(sVramPoolA.Allocate(chunk, size, shift))
            bank &= VRAM_A;
        else
            bank &= ~VRAM_A;
    }

    if(bank & VRAM_B)
    {
        if(sVramPoolB.Allocate(chunk, size, shift))
            bank &= VRAM_B;
        else
            bank &= ~VRAM_B;
    }

    if(!(bank & VRAM_AB))
        return nullptr;

File: libctru/source/allocator/vram.cpp [hunk 5]
```diff
@@ -8,21 +8,28 @@ extern "C"
...
+			return true;
+		}
+		free(blkA);
 	}
 	return false;
 }
 
-void* vramMemAlign(size_t size, size_t alignment)
+void* vramBankMemAlign(VRAM_ALLOCATOR bank, size_t size, size_t alignment)
 {
```
Comment (mtheall) [inline libctru/source/allocator/vram.cpp]: if(bank & ~VRAM_AB)
        return nullptr;

File: libctru/source/allocator/vram.cpp [hunk 6]
```diff
@@ -74,13 +119,30 @@ void vramFree(void* mem)
...
+	if(bank != VRAM_B)
+	{
+		u32 space = sVramPoolA.GetFreeSpace();
+		if(bank == VRAM_A)
+			return space;
+		else
+			return space + sVramPoolB.GetFreeSpace();
+	}
+	else
+		return sVramPoolB.GetFreeSpace();
```
Comment (mtheall) [inline libctru/source/allocator/vram.cpp]: u32 space = 0;

    if(bank & VRAM_A)
        space += sVramPoolA.GetFreeSpace();

    if(bank & VRAM_B)
        space += sVramPoolB.GetFreeSpace();

    return space;

File: libctru/source/allocator/vram.cpp [hunk 7]
```diff
@@ -74,13 +119,30 @@ void vramFree(void* mem)
...
+}
+
 u32 vramSpaceFree()
 {
-	return sVramPool.GetFreeSpace();
+	return sVramPoolA.GetFreeSpace() + sVramPoolB.GetFreeSpace();
```
Comment (mtheall) [inline libctru/source/allocator/vram.cpp]: return vramBankSpaceFree(VRAM_AB);

File: libctru/include/3ds/allocator/vram.h (hunk 1 above)
Comment (fincs) [inline libctru/include/3ds/allocator/vram.h]: How about `VRAM_BANK` instead of `VRAM_ALLOCATOR`?

File: libctru/source/allocator/vram.cpp (hunk 5 above)
Comment (fincs) [inline libctru/source/allocator/vram.cpp]: Can the old `vramMemAlign` be restored, working in a similar way to `vramAlloc`?

File: libctru/source/allocator/vram.cpp [hunk 8]
```diff
@@ -39,27 +46,65 @@ void* vramMemAlign(size_t size, size_t alignment)
...
+{
+	return vramBankMemAlign(VRAM_AB, size, alignment);
+}
+
+void* vramBankAlloc(VRAM_ALLOCATOR bank, size_t size)
+{
+	return vramBankMemAlign(bank, size, 0x80);
+}
+
 void* vramAlloc(size_t size)
```
Comment (fincs) [inline libctru/source/allocator/vram.cpp]: Should probably be turned `static inline`.

**Comment by DiscostewSM** [comment]
I'll admit that I may have rushed into this a little too quickly (if me making silly mistakes in the coding wasn't an obvious sign). I will try and re-create the scenarios I described when I return home today for confirmation on if it's truly a problem.

**Comment by DiscostewSM** [comment]
Hmm, just did a test regarding textures that spanned over both banks (after implementing the suggestions here), and had no issue. Went back to where I read about it, and I guess I misunderstood it. Didn't specifically talk about individual textures, but was mentioning general cube-mapping. Whether that changes anything, I don't know. Have yet to test the other issues as I'm doing some adjustments with citro3d to allow it their testing.

On another note, I wonder. Would it be bad to combine the functions of FCRAM and VRAM allocation together in the kind of manner I attempted to do with the VRAM banks? They seem to basically be identical, though my only issue revolves around the rbtree stuff, as I'm not sure about their operation.

**Comment by fincs** [comment]
For your information, a big update to citro3d's texturing API is currently in the works. Also, I'd prefer to keep linearmem/VRAM allocation APIs separate.

**Comment by DiscostewSM** [comment]
I'm gonna close this because I just haven't had the time to work on it, plus my initial theories on the VRAM seem to conflict with what's actually going on. Maybe I'll come back to it, but from the look of it, it may not be needed at all.

**Pull request closed without merging**